File: ios/Feature/Library/Sources/LibraryFeature/Application/InMemoryPlaybackQueueBridge.swift
import CorePlayer
import Foundation

/// In-memory реализация очереди воспроизведения (параллель Android `InMemoryPlaybackQueueBridge`).
public final class InMemoryPlaybackQueueBridge: PlaybackQueueBridge, @unchecked Sendable {
    private let storage = PlaybackQueueStorage()

    public init() {}

    public func playbackStateStream() -> AsyncStream<PlaybackQueueState> {
        AsyncStream { continuation in
            let id = UUID()
            Task {
                await storage.registerPlaybackContinuation(id: id, continuation: continuation)
            }
        }
    }

    public func stripStateStream() -> AsyncStream<NowPlayingStripExternalState> {
        AsyncStream { continuation in
            let id = UUID()
            Task {
                await storage.registerStripContinuation(id: id, continuation: continuation)
            }
        }
    }

    public func replaceQueue(queue: [PlaybackQueueItem], startIndex: Int?, autoPlay: Bool) async {
        await storage.replaceQueue(queue: queue, startIndex: startIndex, autoPlay: autoPlay)
    }

    public func playTrack(index: Int) async {
        await storage.playTrack(index: index)
    }

    public func play() async {
        await storage.play()
    }

    public func pause() async {
        await storage.pause()
    }

    public func skipNext() async {
        await storage.skipNext()
    }

    public func skipPrevious() async {
        await storage.skipPrevious()
    }

    public func seekTo(positionMs: Int64) async {
        await storage.seekTo(positionMs: positionMs)
    }
}

private actor PlaybackQueueStorage {
    private var playbackState = PlaybackQueueState()
    private var playbackContinuations: [UUID: AsyncStream<PlaybackQueueState>.Continuation] = [:]
    private var stripContinuations: [UUID: AsyncStream<NowPlayingStripExternalState>.Continuation] = [:]

    func registerPlaybackContinuation(
        id: UUID,
        continuation: AsyncStream<PlaybackQueueState>.Continuation
    ) {
        playbackContinuations[id] = continuation
        continuation.yield(playbackState)
        continuation.onTermination = { @Sendable _ in
            Task { await self.removePlaybackContinuation(id: id) }
        }
    }

    func removePlaybackContinuation(id: UUID) {
        playbackContinuations.removeValue(forKey: id)
    }

    func registerStripContinuation(
        id: UUID,
        continuation: AsyncStream<NowPlayingStripExternalState>.Continuation
    ) {
        stripContinuations[id] = continuation
        continuation.yield(stripState(from: playbackState))
        continuation.onTermination = { @Sendable _ in
            Task { await self.removeStripContinuation(id: id) }
        }
    }

    func removeStripContinuation(id: UUID) {
        stripContinuations.removeValue(forKey: id)
    }

    func replaceQueue(queue: [PlaybackQueueItem], startIndex: Int?, autoPlay: Bool) {
        let previousState = playbackState
        let preservedIndex = previousState.currentItem.flatMap { currentItem in
            queue.firstIndex(where: { $0.id == currentItem.id })
        }
        let nextIndex = startIndex ?? preservedIndex
        let currentPositionMs: Int64
        if let nextIndex, nextIndex == preservedIndex,
           nextIndex >= 0, nextIndex < queue.count {
            let durationMs = max(queue[nextIndex].durationMs, 0)
            currentPositionMs = min(
                max(previousState.currentPositionMs, 0),
                durationMs
            )
        } else {
            currentPositionMs = 0
        }

        let isPlaying: Bool
        if queue.isEmpty {
            isPlaying = false
        } else if autoPlay, nextIndex != nil {
            isPlaying = true
        } else if let nextIndex, nextIndex == preservedIndex {
            isPlaying = previousState.isPlaying
        } else {
            isPlaying = false
        }

        playbackState = PlaybackQueueState(
            queue: queue,
            currentIndex: nextIndex,
            isPlaying: isPlaying,
            currentPositionMs: currentPositionMs,
            controlsEnabled: !queue.isEmpty
        )
        broadcast()
    }

    func playTrack(index: Int) {
        guard playbackState.queue.indices.contains(index) else {
            return
        }
        playbackState = PlaybackQueueState(
            queue: playbackState.queue,
            currentIndex: index,
            isPlaying: true,
            currentPositionMs: 0,
            controlsEnabled: !playbackState.queue.isEmpty
        )
        broadcast()
    }

    func play() {
        guard !playbackState.queue.isEmpty else {
            return
        }
        let index = playbackState.currentIndex ?? 0
        playbackState = PlaybackQueueState(
            queue: playbackState.queue,
            currentIndex: index,
            isPlaying: true,
            controlsEnabled: true
        )
        broadcast()
    }

    func pause() {
        guard !playbackState.queue.isEmpty else {
            return
        }
        playbackState = PlaybackQueueState(
            queue: playbackState.queue,
            currentIndex: playbackState.currentIndex,
            isPlaying: false,
            currentPositionMs: playbackState.currentPositionMs,
            controlsEnabled: true
        )
        broadcast()
    }

    func skipNext() {
        guard let currentIndex = playbackState.currentIndex else {
            return
        }
        let nextIndex = min(currentIndex + 1, playbackState.queue.count - 1)
        playbackState = PlaybackQueueState(
            queue: playbackState.queue,
            currentIndex: nextIndex,
            isPlaying: true,
            currentPositionMs: 0,
            controlsEnabled: true
        )
        broadcast()
    }

    func skipPrevious() {
        guard let currentIndex = playbackState.currentIndex else {
            return
        }
        if playbackState.currentPositionMs > restartThresholdMs {
            playbackState = PlaybackQueueState(
                queue: playbackState.queue,
                currentIndex: currentIndex,
                isPlaying: playbackState.isPlaying,
                currentPositionMs: 0,
                controlsEnabled: true
            )
            broadcast()
            return
        }
        let previousIndex = max(currentIndex - 1, 0)
        playbackState = PlaybackQueueState(
            queue: playbackState.queue,
            currentIndex: previousIndex,
            isPlaying: true,
            currentPositionMs: 0,
            controlsEnabled: true
        )
        broadcast()
    }

    func seekTo(positionMs: Int64) {
        guard let currentItem = playbackState.currentItem else {
            return
        }
        let maxMs = max(currentItem.durationMs, 0)
        let clamped = min(max(positionMs, 0), maxMs)
        playbackState = PlaybackQueueState(
            queue: playbackState.queue,
            currentIndex: playbackState.currentIndex,
            isPlaying: playbackState.isPlaying,
            currentPositionMs: clamped,
            controlsEnabled: true
        )
        broadcast()
    }

    private func broadcast() {
        let state = playbackState
        let strip = stripState(from: state)
        for continuation in playbackContinuations.values {
            continuation.yield(state)
        }
        for continuation in stripContinuations.values {
            continuation.yield(strip)
        }
    }

    private func stripState(from playback: PlaybackQueueState) -> NowPlayingStripExternalState {
        let current = playback.currentItem
        return NowPlayingStripExternalState(
            title: current?.title ?? "",
            subtitle: current?.subtitle ?? "",
            isPlaying: playback.isPlaying,
            currentPositionMs: playback.currentPositionMs,
            durationMs: current.map { max($0.durationMs, 0) } ?? 0,
            controlsEnabled: playback.controlsEnabled
        )
    }
}

private let restartThresholdMs: Int64 = 3_000

File: android/feature/library/src/main/java/com/mplayeraudio/feature/library/InMemoryPlaybackQueueBridge.kt
package com.mplayeraudio.feature.library

import com.mplayeraudio.core.player.NowPlayingStripExternalState
import com.mplayeraudio.core.player.PlaybackPhase
import com.mplayeraudio.core.player.PlaybackQueueBridge
import com.mplayeraudio.core.player.PlaybackQueueItem
import com.mplayeraudio.core.player.PlaybackQueueState
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.map

internal class InMemoryPlaybackQueueBridge : PlaybackQueueBridge {

    private val stateFlow = MutableStateFlow(PlaybackQueueState())

    override val playbackState: StateFlow<PlaybackQueueState> = stateFlow.asStateFlow()

    override val state: Flow<NowPlayingStripExternalState> = playbackState.map { ps ->
        val currentItem = ps.currentItem
        NowPlayingStripExternalState(
            title = currentItem?.title.orEmpty(),
            subtitle = currentItem?.subtitle.orEmpty(),
            isPlaying = ps.isPlaying,
            currentPositionMs = ps.currentPositionMs,
            durationMs = ps.currentDurationMs ?: currentItem?.durationMs ?: 0L,
            controlsEnabled = ps.controlsEnabled,
        )
    }

    override suspend fun replaceQueue(
        queue: List<PlaybackQueueItem>,
        startIndex: Int?,
        autoPlay: Boolean,
    ) {
        val previous = stateFlow.value
        val preservedIndex = previous.currentItem
            ?.let { curr -> queue.indexOfFirst { it.id == curr.id }.takeIf { it >= 0 } }
        val nextIndex = startIndex ?: preservedIndex
        val preservingCurrentTrack = nextIndex != null && nextIndex == preservedIndex
        val currentPositionMs = if (preservingCurrentTrack && nextIndex != null) {
            previous.currentPositionMs.coerceIn(0L, queue[nextIndex].durationMs.coerceAtLeast(0L))
        } else {
            0L
        }
        val phase = when {
            queue.isEmpty() -> PlaybackPhase.Idle
            autoPlay && nextIndex != null -> PlaybackPhase.Playing
            preservingCurrentTrack -> previous.phase
            else -> PlaybackPhase.Idle
        }

        stateFlow.value = PlaybackQueueState(
            queue = queue,
            currentIndex = nextIndex,
            phase = phase,
            currentPositionMs = currentPositionMs,
        )
    }

    override suspend fun playTrack(index: Int) {
        val current = stateFlow.value
        if (index !in current.queue.indices) return
        stateFlow.value = current.copy(
            currentIndex = index,
            phase = PlaybackPhase.Playing,
            currentPositionMs = 0L,
        )
    }

    override suspend fun play() {
        val current = stateFlow.value
        if (current.queue.isEmpty()) return
        stateFlow.value = current.copy(
            currentIndex = current.currentIndex ?: 0,
            phase = PlaybackPhase.Playing,
        )
    }

    override suspend fun pause() {
        val current = stateFlow.value
        if (current.queue.isEmpty()) return
        stateFlow.value = current.copy(phase = PlaybackPhase.Paused)
    }

    override suspend fun skipNext() {
        val current = stateFlow.value
        val currentIndex = current.currentIndex ?: return
        val nextIndex = (currentIndex + 1).coerceAtMost(current.queue.lastIndex)
        stateFlow.value = current.copy(
            currentIndex = nextIndex,
            currentPositionMs = 0L,
            phase = PlaybackPhase.Playing,
        )
    }

    override suspend fun skipPrevious() {
        val current = stateFlow.value
        val currentIndex = current.currentIndex ?: return
        if (current.currentPositionMs > RestartThresholdMs) {
            stateFlow.value = current.copy(currentPositionMs = 0L)
            return
        }
        val previousIndex = (currentIndex - 1).coerceAtLeast(0)
        stateFlow.value = current.copy(
            currentIndex = previousIndex,
            currentPositionMs = 0L,
            phase = PlaybackPhase.Playing,
        )
    }

    override suspend fun seekTo(positionMs: Long) {
        val current = stateFlow.value
        val currentItem = current.currentItem ?: return
        stateFlow.value = current.copy(
            currentPositionMs = positionMs.coerceIn(0L, currentItem.durationMs.coerceAtLeast(0L)),
        )
    }

    override fun acknowledgeError() {
        stateFlow.value = stateFlow.value.copy(playbackError = null)
    }

    override fun shutdown() {
        stateFlow.value = PlaybackQueueState()
    }
}

private const val RestartThresholdMs = 3_000L
